src/Databases/DatabaseOnDisk.cpp

```diff
@@ -213,8 +218,47 @@ void DatabaseOnDisk::commitCreateTable(const ASTCreateQuery & query, const Stora
         Poco::File(table_metadata_tmp_path).remove();
         throw;
     }
+
+    try
+    {
+        /// If the table was detached permanently we will have a file with
+        /// .sql_detached suffix, which is not needed anymore since we attached the table back
+        auto table_metadata_file_detached = Poco::File(table_metadata_path + detached_suffix);
+        if (table_metadata_file_detached.exists())
+            table_metadata_file_detached.remove();
+    }
+    catch (...)
+    {
+        // It's not a big issue if we can't remove the .sql_detached file.
+        LOG_WARNING(log, getCurrentExceptionMessage(__PRETTY_FUNCTION__));
+    }
+}
+
+void DatabaseOnDisk::detachTablePermanently(const String & table_name)
+{
+    StoragePtr table = detachTable(table_name);
+
+    /// This is possible for Lazy database.
+    if (!table)
+        return;
+
+    String table_metadata_path = getObjectMetadataPath(table_name);
+    String table_metadata_path_detached = table_metadata_path + detached_suffix;
+
+    try
+    {
+        /// it will silently overwrite the file if exists, and it's ok
```

Comment:
Maybe use `renameNoReplace(...)`?

Reply:
Actually, the `sql_detached` should not exist in normal circumstances. And if it was existing I guess the best what we can do is just overwrite it (it's either the same, either some obsolete version of the schema). From what I've understood  `renameNoReplace` does a bit different thing.

Reply:
```
create table test (n int) engine=MergeTree order by n;
detach table test permanently; 
create table test2 (s String) engine=Log;
rename table test2 to test;

-- overwrites previous test.sql_detached
detach table test permanently;
```

Reply:
Even in that case - we can't reattach the original table back (it's name was taken by newer table), so keeping the original .sql_detached file will be not very useful.

added a checks to prevent reusing the table name while it's detached, I think that should be ok now (will add your test).